diff -Nuar kernel/kernel-jammy-src/drivers/media/v4l2-core/v4l2-ioctl.c kernel_new/kernel-jammy-src/drivers/media/v4l2-core/v4l2-ioctl.c
--- kernel/kernel-jammy-src/drivers/media/v4l2-core/v4l2-ioctl.c	2024-07-23 19:15:48.538330128 -0700
+++ kernel_new/kernel-jammy-src/drivers/media/v4l2-core/v4l2-ioctl.c	2024-08-07 18:21:32.905351839 -0700
@@ -1378,6 +1378,12 @@
 	case V4L2_PIX_FMT_TM6000:	descr = "A/V + VBI Mux Packet"; break;
 	case V4L2_PIX_FMT_CIT_YYVYUY:	descr = "GSPCA CIT YYVYUY"; break;
 	case V4L2_PIX_FMT_KONICA420:	descr = "GSPCA KONICA420"; break;
+    //xumm add
+    case V4L2_PIX_FMT_TX2_Y10:	descr = "TX2 10-bit/16-bit Greyscale"; break;
+	case V4L2_PIX_FMT_TX2_Y12:	descr = "TX2 12-bit/16-bit Greyscale"; break;
+	case V4L2_PIX_FMT_XAVIER_Y10:	descr = "XAVIER 10-bit/16-bit Greyscale"; break;
+	case V4L2_PIX_FMT_XAVIER_Y12:	descr = "XAVIER 12-bit/16-bit Greyscale"; break;
+    //end
 	case V4L2_PIX_FMT_HSV24:	descr = "24-bit HSV 8-8-8"; break;
 	case V4L2_PIX_FMT_HSV32:	descr = "32-bit XHSV 8-8-8-8"; break;
 	case V4L2_SDR_FMT_CU8:		descr = "Complex U8"; break;
diff -Nuar kernel/kernel-jammy-src/include/uapi/linux/videodev2.h kernel_new/kernel-jammy-src/include/uapi/linux/videodev2.h
--- kernel/kernel-jammy-src/include/uapi/linux/videodev2.h	2024-07-23 19:15:51.798292569 -0700
+++ kernel_new/kernel-jammy-src/include/uapi/linux/videodev2.h	2024-08-07 18:21:53.908300183 -0700
@@ -748,6 +748,15 @@
 #define V4L2_PIX_FMT_IPU3_SGBRG10	v4l2_fourcc('i', 'p', '3', 'g') /* IPU3 packed 10-bit GBRG bayer */
 #define V4L2_PIX_FMT_IPU3_SGRBG10	v4l2_fourcc('i', 'p', '3', 'G') /* IPU3 packed 10-bit GRBG bayer */
 #define V4L2_PIX_FMT_IPU3_SRGGB10	v4l2_fourcc('i', 'p', '3', 'r') /* IPU3 packed 10-bit RGGB bayer */
+//xumm add
+/* TX2 */
+#define V4L2_PIX_FMT_TX2_Y10     v4l2_fourcc('T', 'Y', '1', '0') /* 10  Greyscale     */
+#define V4L2_PIX_FMT_TX2_Y12     v4l2_fourcc('T', 'Y', '1', '2') /* 12  Greyscale     */
+
+/* Xavier */
+#define V4L2_PIX_FMT_XAVIER_Y10     v4l2_fourcc('X', 'Y', '1', '0') /* 10  Greyscale     */
+#define V4L2_PIX_FMT_XAVIER_Y12     v4l2_fourcc('X', 'Y', '1', '2') /* 12  Greyscale     */
+//end
 
 /* SDR formats - used only for Software Defined Radio devices */
 #define V4L2_SDR_FMT_CU8          v4l2_fourcc('C', 'U', '0', '8') /* IQ u8 */
diff -Nuar nvidia-oot/drivers/media/platform/tegra/camera/camera_common.c nvidia-oot_new/drivers/media/platform/tegra/camera/camera_common.c
--- nvidia-oot/drivers/media/platform/tegra/camera/camera_common.c	2024-07-23 19:16:11.746062795 -0700
+++ nvidia-oot_new/drivers/media/platform/tegra/camera/camera_common.c	2024-08-07 18:23:57.890949519 -0700
@@ -89,6 +89,26 @@
 		V4L2_COLORSPACE_SRGB,
 		V4L2_PIX_FMT_VYUY,
 	},
+	//xumm add
+	// Grayscale 8bit support
+	{
+		MEDIA_BUS_FMT_Y8_1X8,
+		V4L2_COLORSPACE_RAW,
+		V4L2_PIX_FMT_GREY,
+	},
+	// Grayscale 10bit support
+	{
+		MEDIA_BUS_FMT_Y10_1X10,
+		V4L2_COLORSPACE_RAW,
+		V4L2_PIX_FMT_Y10,
+	},
+	// Grayscale 12bit support
+	{
+		MEDIA_BUS_FMT_Y12_1X12,
+		V4L2_COLORSPACE_RAW,
+		V4L2_PIX_FMT_Y12,
+	},
+	//xumm add end
 	{
 		MEDIA_BUS_FMT_RGB888_1X24,
 		V4L2_COLORSPACE_SRGB,
diff -Nuar nvidia-oot/drivers/media/platform/tegra/camera/sensor_common.c nvidia-oot_new/drivers/media/platform/tegra/camera/sensor_common.c
--- nvidia-oot/drivers/media/platform/tegra/camera/sensor_common.c	2024-07-23 19:16:11.746062795 -0700
+++ nvidia-oot_new/drivers/media/platform/tegra/camera/sensor_common.c	2024-08-07 18:24:03.260786080 -0700
@@ -267,6 +267,14 @@
 		*format = V4L2_PIX_FMT_UYVY;
 	else if (strncmp(pixel_t, "yuv_vyuy16", size) == 0)
 		*format = V4L2_PIX_FMT_VYUY;
+    //xumm add
+    else if (strncmp(pixel_t, "raw_y88", size) == 0)
+         *format = V4L2_PIX_FMT_GREY; //Grayscale
+    else if (strncmp(pixel_t, "raw_y1010", size) == 0)
+         *format = V4L2_PIX_FMT_Y10; //
+    else if (strncmp(pixel_t, "raw_y1212", size) == 0)
+         *format = V4L2_PIX_FMT_Y12; //
+    //xum add end
 	else {
 		pr_err("%s: Need to extend format%s\n", __func__, pixel_t);
 		return -EINVAL;
diff -Nuar nvidia-oot/drivers/media/platform/tegra/camera/vi/channel.c nvidia-oot_new/drivers/media/platform/tegra/camera/vi/channel.c
--- nvidia-oot/drivers/media/platform/tegra/camera/vi/channel.c	2024-07-23 19:16:11.746062795 -0700
+++ nvidia-oot_new/drivers/media/platform/tegra/camera/vi/channel.c	2024-08-07 18:23:32.325045158 -0700
@@ -1304,6 +1304,13 @@
 	case TEGRA_CAMERA_CID_LOW_LATENCY:
 		chan->low_latency = ctrl->val;
 		break;
+//xumm add
+    case TEGRA_CAMERA_CID_VI_TIME_OUT_DISABLE:
+        chan->vi_timeout_disable = ctrl->val;
+        dev_info(&chan->video->dev, "%s: %u set vi timeout disable %d\n",
+			__func__, ctrl->id,ctrl->val);
+        break;
+//add end
 	case TEGRA_CAMERA_CID_VI_PREFERRED_STRIDE:
 		chan->preferred_stride = ctrl->val;
 		tegra_channel_update_format(chan, chan->format.width,
@@ -1477,6 +1484,18 @@
 		.step = 1,
 		.def = 0,
 	},
+	//xumm add
+	{
+		.ops = &channel_ctrl_ops,
+		.id = TEGRA_CAMERA_CID_VI_TIME_OUT_DISABLE,
+		.name = "VI Time Out Disable",
+		.type = V4L2_CTRL_TYPE_BOOLEAN,
+		.def = 0,
+		.min = 0,
+		.max = 1,
+		.step = 1,
+	},
+	//add end
 };
 
 #define GET_TEGRA_CAMERA_CTRL(id, c)					\
diff -Nuar nvidia-oot/drivers/media/platform/tegra/camera/vi/vi5_fops.c nvidia-oot_new/drivers/media/platform/tegra/camera/vi/vi5_fops.c
--- nvidia-oot/drivers/media/platform/tegra/camera/vi/vi5_fops.c	2024-07-23 19:16:11.750062749 -0700
+++ nvidia-oot_new/drivers/media/platform/tegra/camera/vi/vi5_fops.c	2024-08-07 18:23:41.987018102 -0700
@@ -507,10 +507,20 @@
 			goto rel_buf;
 
 		/* Dequeue a frame and check its capture status */
-		err = vi_capture_status(chan->tegra_vi_channel[vi_port], CAPTURE_TIMEOUT_MS);
-		if (err) {
-			if (err == -ETIMEDOUT) {
-				dev_err(vi->dev,
+		//xumm add
+		//err = vi_capture_status(chan->tegra_vi_channel[vi_port], CAPTURE_TIMEOUT_MS);
+		do{
+    		err = vi_capture_status(chan->tegra_vi_channel[vi_port], CAPTURE_TIMEOUT_MS);
+    		if (err) {
+                //xumm 
+                if (err == -ETIMEDOUT && chan->vi_timeout_disable) {
+                    dev_err(vi->dev,
+    					"xavier capture time out , will wait\n");
+                    continue;
+                }
+                //end
+    			else if (err == -ETIMEDOUT) {
+					dev_err(vi->dev,
 					"uncorr_err: request timed out after %d ms\n",
 					CAPTURE_TIMEOUT_MS);
 			} else {
@@ -539,6 +549,10 @@
 			goto uncorr_err;
 		}
 
+        //xumm add ,this means get a frame 
+        	break;
+        }while(chan->vi_timeout_disable);
+
 		spin_lock_irqsave(&chan->capture_state_lock, flags);
 		if (chan->capture_state != CAPTURE_ERROR) {
 			chan->capture_reqs_enqueued -= 1;
diff -Nuar nvidia-oot/drivers/media/platform/tegra/camera/vi/vi5_formats.h nvidia-oot_new/drivers/media/platform/tegra/camera/vi/vi5_formats.h
--- nvidia-oot/drivers/media/platform/tegra/camera/vi/vi5_formats.h	2024-07-23 19:16:11.750062749 -0700
+++ nvidia-oot_new/drivers/media/platform/tegra/camera/vi/vi5_formats.h	2024-08-07 18:23:48.029626933 -0700
@@ -136,6 +136,14 @@
 				YUV422_8, YUYV, "YUV 4:2:2 YUYV"),
 	TEGRA_VIDEO_FORMAT(YUV422, 16, YVYU8_2X8, 2, 1, T_Y8_V8__Y8_U8,
 				YUV422_8, YVYU, "YUV 4:2:2 YVYU"),
+    //xumm add
+    TEGRA_VIDEO_FORMAT(RAW8, 8, Y8_1X8, 1, 1, T_R8,
+				RAW8, GREY, "GREY8"),
+    TEGRA_VIDEO_FORMAT(RAW10, 10, Y10_1X10, 2, 1, T_R16,
+				RAW10, XAVIER_Y10, "GREY10"),
+	TEGRA_VIDEO_FORMAT(RAW12, 12, Y12_1X12, 2, 1, T_R16,
+				RAW12, XAVIER_Y12, "GREY12"),
+	//xumm add end
 };
 
 #endif
diff -Nuar nvidia-oot/include/media/mc_common.h nvidia-oot_new/include/media/mc_common.h
--- nvidia-oot/include/media/mc_common.h	2024-07-23 19:16:11.930060676 -0700
+++ nvidia-oot_new/include/media/mc_common.h	2024-08-07 18:24:29.743055518 -0700
@@ -264,6 +264,9 @@
 	dma_addr_t emb_buf;
 	void *emb_buf_addr;
 	unsigned int emb_buf_size;
+    //xumm add
+    bool vi_timeout_disable;
+    //end
 };
 
 #define to_tegra_channel(vdev) \
diff -Nuar nvidia-oot/include/media/tegra-v4l2-camera.h nvidia-oot_new/include/media/tegra-v4l2-camera.h
--- nvidia-oot/include/media/tegra-v4l2-camera.h	2024-07-23 19:16:11.930060676 -0700
+++ nvidia-oot_new/include/media/tegra-v4l2-camera.h	2024-08-07 18:24:21.768550681 -0700
@@ -48,6 +48,8 @@
 #define TEGRA_CAMERA_CID_SENSOR_DV_TIMINGS         (TEGRA_CAMERA_CID_BASE+108)
 #define TEGRA_CAMERA_CID_LOW_LATENCY         (TEGRA_CAMERA_CID_BASE+109)
 #define TEGRA_CAMERA_CID_VI_PREFERRED_STRIDE (TEGRA_CAMERA_CID_BASE+110)
+//xumm add
+#define TEGRA_CAMERA_CID_VI_TIME_OUT_DISABLE (TEGRA_CAMERA_CID_BASE+120)
 
 /**
  * This is temporary with the current v4l2 infrastructure
